src/utils/DependencyUtils.ts

```diff
@@ -11,3 +29,71 @@ export const splitToNameAndVersion = (dependencyString: string): [string, string
     const fullName = parts.join('-');
     return [fullName, version];
 };
+
+/**
+ * - When installing a regular mods, always install the latest versions of dependencies.
+ * - When installing a mod pack we usually want to ensure compatibility by installing the exact
+ *      versions of dependencies, even if it means upgrading/downgrading already installed mods.
+ *   - If, however, modpackDependencyStrategy is set to USE_LATEST_VERSION, we install the latest
+ *      versions of dependencies for mod packs too.
+ * - When installing latest versions of dependencies, already installed mods are filtered out.
+ * - The order of the input combos array matters, combos earlier on the array get their dependencies'
+ *      versions prioritised over later ones.
+ */
+export async function getFullDependencyList(
+    combos: ThunderstoreCombo[],
+    game: Game,
+    installedMods: ManifestV2[],
+    modpackDependencyStrategy: ModpackDependencyStrategy
+) {
+    const originalCombos = [...combos];
+    for (const combo of originalCombos) {
+        const isModpack = combo.getMod().getCategories().some(value => value === "Modpacks");
+        const dependencyVersionStrategy = isModpack ? modpackDependencyStrategy : ModpackDependencyStrategy.USE_LATEST_VERSION;
+
+        let dependencies: ThunderstoreCombo[] = [];
+        await buildDependencySet(combo.getVersion(), game, dependencies, dependencyVersionStrategy);
+
+        combos.push(...dependencies.filter(dependency => {
+            const isNotAlreadyFound = !combos.some(alreadyAdded => alreadyAdded.getMod().getFullName() === dependency.getMod().getFullName());
+            const isNotInstalled = !installedMods.some(installed => installed.getName() === dependency.getMod().getFullName());
+
+            return isNotAlreadyFound && (isNotInstalled || dependencyVersionStrategy === ModpackDependencyStrategy.USE_EXACT_VERSION);
```

Comment:
This combination of things is a bit hard to grasp and e.g. maintain. That, and the subtle changes in the implementation is probably due to the previous implementation having two separate methods that a bit arbitrary did things differently. Combining them into one function can easily go awry. While spending way too much time thinking about this, I think I came up with a solution that is a bit easier to understand and maintain while being closer to the original functionality.

(Another solution would've been to split "install one" and "update all" to separate functions but then we'd be pretty close to square 1 again.)